### ISSUE TITLE
Identify operator selection policy by Atlas

### ISSUE THREAD

**Issue by dobertRowneySr** [Joystream/joystream#5027]
SSIA.
It's not clear yet how exactly Atlas identifies distributor nodes, especially when geolocation data is missing

**Comment by dobertRowneySr** [comment]
# Video Player
**scope**: detail how videos are fetched in atlas in the video player page view, that is once I am on the home page and then I click on any of the video suggested

## Stack Trace
1.[`GetFullVideo` query](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/api/queries/videos.graphql#L87) is called. The [fragment](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/api/queries/fragments.graphql#L176) contains several fields among which a [`media.resolvedUrls`](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/api/queries/fragments.graphql#L101). 
2. The distributor node url list is put together in this function [here](https://github.com/ignazio-bovo/orion/blob/0ae1af5f11f3e3414acd85dfedfa0bd9a76af330/src/server-extension/resolvers/AssetsResolver/utils.ts#L13) and they are sorted by user geographical proximity if user location data is available, otherwise they are randomly sorted. `userLocation` available in the Application tab > Local Storage > `gleev.xyz` on developer tools of most browsers
3. The `resolvedUrls` are passed to the `VideoComponents` and eventually passed to [`useGetAssetUrl`](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/hooks/useGetAssetUrl.ts#L80). The part where the logic select the right distributor node + asset for the react component is where the [`resolvedUrl` variable is initialized](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/hooks/useGetAssetUrl.ts#L93) via `useSingleAssetUrl`
4. the `useSingleAssetUrl` runs orderly through every distribution url and for each construct a `testAssetDownload` promise. Subsequently select the first url that [resolves that promise before a specific `timeout`](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/utils/misc.ts#L14). If no url is able to satisfy the timeout time constraint then it select the first url that resolve the promise constructed previously. the `testAssetDownload` when the asset is of type video if probes the url by[ sending a `HEAD` request and resolving when the first byte is received](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/providers/assets/assets.helpers.ts#L77)

## TLDR Distribution url selection for videos
1. URLs for a video are sorted either by geographical proximity to the user if geolocation is provided otherwise they are randomly sorted
2. The final url for the streaming is selected as the closest (in the geographical sense) whose sampled time to first byte is less than a specified `timeOut`. If none of the url is able to provide a time to first byte lower than the `timeOut` value then   the distribution node that is able to provide the lowest sampled time to first byte

# Thumbnails
**scope**: detail how video thumbnails are fetched in atlas
Since for each asset a sequence of `resolvedUrl` is similarly fetched, the process for image asset is identical in terms of DN urls choice

**Comment by kdembler** [comment]
@ignazio-bovo I think we still need to look into those timeouts logic. I have seen problematic behaviour on slow connections where Atlas kept switching provider, not giving any single URL enough time to load. This most likely resulted in worse UX if it just waited a bit longer for the first URL to load

**Comment by dobertRowneySr** [comment]
From the source code if none of the url is resolved within a specified timeout, then [the first url to resolve is used](https://github.com/Joystream/atlas/blob/0efb8ced2014293f2e84e1c23a36fbd73cac1d9a/packages/atlas/src/hooks/useGetAssetUrl.ts#L61)
The timeout value default value is [here](https://github.com/Joystream/atlas/blob/35d064dc7c245e952c05991a02907cd15c847996/packages/atlas/atlas.config.yml#L15C1-L16C1)

**Comment by kdembler** [comment]
Yeah but from my experience on bad connection, the logic could be improved. If let's say video needs 500ms to start playing, but timeout gets set at 400ms, then Atlas will try X URLs for 400ms resulting in `X * 400ms` wait for the user, but if it just waited a bit longer, it would be only 500ms.
